api/src/main/kotlin/com/github/jengelman/gradle/plugins/shadow/tasks/ShadowCopyAction.kt

```diff
@@ -0,0 +1,447 @@
+
+open class ShadowCopyAction internal constructor(
+    private val zipFile: File,
+    private val compressor: ZipCompressor,
+    private val documentationRegistry: DocumentationRegistry,
+    private val encoding: String?,
+    private val transformers: List<Transformer>,
+    private val relocators: List<Relocator>,
+    private val patternSet: PatternSet,
+    private val stats: ShadowStats,
+    private val preserveFileTimestamps: Boolean,
+    private val minimizeJar: Boolean,
+    private val unusedTracker: UnusedTracker?,
+) : CopyAction {
+
+    constructor(
+        zipFile: File,
+        compressor: ZipCompressor,
+        documentationRegistry: DocumentationRegistry,
+        encoding: String?,
+        transformers: List<Transformer>,
+        relocators: List<Relocator>,
+        patternSet: PatternSet,
+        stats: ShadowStats,
+        preserveFileTimestamps: Boolean,
+        minimizeJar: Boolean,
+    ) : this(
+        zipFile,
+        compressor,
+        documentationRegistry,
+        encoding,
+        transformers,
+        relocators,
+        patternSet,
+        stats,
+        preserveFileTimestamps,
+        minimizeJar,
+        null,
+    )
+
+    override fun execute(stream: CopyActionProcessingStream): WorkResult {
+        val unusedClasses = if (minimizeJar && unusedTracker != null) {
+            stream.process(
+                object : BaseStreamAction() {
+                    override fun visitFile(fileDetails: FileCopyDetails) {
+                        // All project sources are already present, we just need
+                        // to deal with JAR dependencies.
+                        if (isArchive(fileDetails)) {
+                            unusedTracker.addDependency(fileDetails.file)
+                        }
+                    }
+                },
+            )
+            unusedTracker.findUnused()
+        } else {
+            emptySet()
+        }
+
+        val zipOutStream = try {
+            compressor.createArchiveOutputStream(zipFile) as ZipOutputStream
+        } catch (e: Exception) {
+            throw GradleException("Could not create ZIP '$zipFile'", e)
+        }
+
+        try {
+            zipOutStream.use { outputStream ->
+                stream.process(
+                    StreamAction(
+                        outputStream,
+                        encoding,
+                        transformers,
+                        relocators,
+                        patternSet,
+                        unusedClasses,
+                        stats,
+                    ),
+                )
+                processTransformers(outputStream)
+            }
+        } catch (e: IOException) {
+            throw Zip64RequiredException(
+                "${e.cause?.message}\n\nTo build this archive, please enable the zip64 extension.\n" +
+                    "See: ${documentationRegistry.getDslRefForProperty(Zip::class.java, "zip64")}",
+            )
```

Comment:
We shouldn't rethrow all `IOException`s to `Zip64RequiredException`s, fixing in #1147.